Cannot find module 'native-reg' on Linux

<!--
  Hi there! Thank you for discovering and submitting an issue.

  Before you submit this; let's make sure of a few things.
  Please make sure the following boxes are ticked if they are correct.
  If not, please try and fulfill these first.
-->

<!--
  Once those are done, if you're able to fill in the following list with your information,
  it'd be very helpful to whoever handles the issue.
-->

- **OS version and name**: Arch Linux
- **Hyper.app version**: 3.1.0-canary.4

## Issue
<!-- Now feel free to write your issue, but please be descriptive! Thanks again 🙌 ❤️ -->
```
❯ ./Hyper-3.1.0-canary.4.AppImage
Error: Cannot find module 'native-reg'
Require stack:
- /tmp/.mount_Hyper-Rmq6V0/resources/app.asar/utils/cli-install.js
- /tmp/.mount_Hyper-Rmq6V0/resources/app.asar/index.js
- 
    at Module._resolveFilename (internal/modules/cjs/loader.js:717:15)
    at Function../lib/common/reset-search-paths.ts.Module._resolveFilename (electron/js2c/browser_init.js:7646:16)
    at Module._load (internal/modules/cjs/loader.js:622:27)
    at Module._load (electron/js2c/asar.js:717:26)
    at Function.Module._load (electron/js2c/asar.js:717:26)
    at Function._load (/tmp/.mount_Hyper-Rmq6V0/resources/app.asar/plugins.js:82:37)
    at Module.require (internal/modules/cjs/loader.js:775:19)
    at require (internal/modules/cjs/helpers.js:68:18)
    at Object.<anonymous> (/tmp/.mount_Hyper-Rmq6V0/resources/app.asar/utils/cli-install.js:13:20)
    at Module._compile (internal/modules/cjs/loader.js:880:30) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/tmp/.mount_Hyper-Rmq6V0/resources/app.asar/utils/cli-install.js',
    '/tmp/.mount_Hyper-Rmq6V0/resources/app.asar/index.js',
    undefined
  ]
}
```

--- a/app/config/open.ts
+++ b/app/config/open.ts
@@ -1,65 +1,74 @@
 import {shell} from 'electron';
 import {cfgPath} from './paths';
-export default () => Promise.resolve(shell.openItem(cfgPath));
+import * as regTypes from '../typings/native-reg';
 
-// Windows opens .js files with  WScript.exe by default
-// If the user hasn't set up an editor for .js files, we fallback to notepad.
-if (process.platform === 'win32') {
-  const Registry = require('winreg') as typeof import('winreg');
-  const {exec} = require('child_process') as typeof import('child_process');
+export default () => {
+  // Windows opens .js files with  WScript.exe by default
+  // If the user hasn't set up an editor for .js files, we fallback to notepad.
+  if (process.platform === 'win32') {
+    try {
+      // eslint-disable-next-line no-var, @typescript-eslint/no-var-requires
+      var Registry: typeof regTypes = require('native-reg');
+    } catch (err) {
+      console.error(err);
+    }
+    const {exec} = require('child_process') as typeof import('child_process');
 
-  const getUserChoiceKey = async () => {
-    // Load FileExts keys for .js files
-    const keys: Winreg.Registry[] = await new Promise((resolve, reject) => {
-      new Registry({
-        hive: Registry.HKCU,
-        key: '\\Software\\Microsoft\\Windows\\CurrentVersion\\Explorer\\FileExts\\.js'
-      }).keys((error, items) => {
-        if (error) {
-          reject(error);
-        } else {
-          resolve(items || []);
-        }
+    const getUserChoiceKey = async () => {
+      try {
+        // Load FileExts keys for .js files
+        const fileExtsKeys = Registry.openKey(
+          Registry.HKCU,
+          'Software\\Microsoft\\Windows\\CurrentVersion\\Explorer\\FileExts\\.js',
+          Registry.Access.READ
+        );
+        const keys = fileExtsKeys ? Registry.enumKeyNames(fileExtsKeys) : [];
+        Registry.closeKey(fileExtsKeys);
+
+        // Find UserChoice key
+        const userChoice = keys.find(k => k.endsWith('UserChoice'));
+        return userChoice
+          ? `Software\\Microsoft\\Windows\\CurrentVersion\\Explorer\\FileExts\\.js\\${userChoice}`
+          : userChoice;
+      } catch (error) {
+        console.error(error);
+        return;
+      }
+    };
+
+    const hasDefaultSet = async () => {
+      const userChoice = await getUserChoiceKey();
+      if (!userChoice) return false;
+
+      try {
+        // Load key values
+        const userChoiceKey = Registry.openKey(Registry.HKCU, userChoice, Registry.Access.READ)!;
+        const values: string[] = Registry.enumValueNames(userChoiceKey).map(
+          x => (Registry.queryValue(userChoiceKey, x) as string) || ''
+        );
+        Registry.closeKey(userChoiceKey);
+
+        // Look for default program
+        const hasDefaultProgramConfigured = values.every(
+          value => value && typeof value === 'string' && !value.includes('WScript.exe') && !value.includes('JSFile')
+        );
+
+        return hasDefaultProgramConfigured;
+      } catch (error) {
+        console.error(error);
+        return false;
+      }
+    };
+
+    // This mimics shell.openItem, true if it worked, false if not.
+    const openNotepad = (file: string) =>
+      new Promise<boolean>(resolve => {
+        exec(`start notepad.exe ${file}`, error => {
+          resolve(!error);
+        });
       });
-    });
 
-    // Find UserChoice key
-    const userChoice = keys.find(k => k.key.endsWith('UserChoice'));
-    return userChoice;
-  };
-
-  const hasDefaultSet = async () => {
-    const userChoice = await getUserChoiceKey();
-    if (!userChoice) return false;
-
-    // Load key values
-    const values: string[] = await new Promise((resolve, reject) => {
-      userChoice.values((error, items) => {
-        if (error) {
-          reject(error);
-        }
-        resolve(items.map(item => item.value || '') || []);
-      });
-    });
-
-    // Look for default program
-    const hasDefaultProgramConfigured = values.every(
-      value => value && typeof value === 'string' && !value.includes('WScript.exe') && !value.includes('JSFile')
-    );
-
-    return hasDefaultProgramConfigured;
-  };
-
-  // This mimics shell.openItem, true if it worked, false if not.
-  const openNotepad = (file: string) =>
-    new Promise(resolve => {
-      exec(`start notepad.exe ${file}`, error => {
-        resolve(!error);
-      });
-    });
-
-  module.exports = () =>
-    hasDefaultSet()
+    return hasDefaultSet()
       .then(yes => {
         if (yes) {
           return shell.openItem(cfgPath);
@@ -71,4 +80,7 @@ if (process.platform === 'win32') {
         console.error('Open config with default app error:', err);
         return openNotepad(cfgPath);
       });
-}
+  } else {
+    return Promise.resolve(shell.openItem(cfgPath));
+  }
+};

--- a/app/package.json
+++ b/app/package.json
@@ -33,10 +33,9 @@
     "react-dom": "16.13.0",
     "semver": "7.1.3",
     "shell-env": "3.0.0",
-    "uuid": "7.0.1",
-    "winreg": "1.2.4"
+    "uuid": "7.0.1"
   },
   "optionalDependencies": {
-    "native-reg": "0.3.3"
+    "native-reg": "0.3.4"
   }
 }

--- a/app/system-context-menu.ts
+++ b/app/system-context-menu.ts
@@ -1,80 +1,61 @@
-import Registry from 'winreg';
+import * as regTypes from './typings/native-reg';
+if (process.platform === 'win32') {
+  try {
+    // eslint-disable-next-line no-var, @typescript-eslint/no-var-requires
+    var Registry: typeof regTypes = require('native-reg');
+  } catch (err) {
+    console.error(err);
+  }
+}
 
 const appPath = `"${process.execPath}"`;
-const regKey = `\\Software\\Classes\\Directory\\background\\shell\\Hyper`;
+const regKey = `Software\\Classes\\Directory\\background\\shell\\Hyper`;
 const regParts = [
   {key: 'command', name: '', value: `${appPath} "%V"`},
   {name: '', value: 'Open Hyper here'},
   {name: 'Icon', value: `${appPath}`}
 ];
 
-function addValues(hyperKey: Registry.Registry, commandKey: Registry.Registry, callback: Function) {
-  hyperKey.set(regParts[1].name, Registry.REG_SZ, regParts[1].value, error => {
-    if (error) {
-      console.error(error.message);
-    }
-    hyperKey.set(regParts[2].name, Registry.REG_SZ, regParts[2].value, err => {
-      if (err) {
-        console.error(err.message);
-      }
-      commandKey.set(regParts[0].name, Registry.REG_SZ, regParts[0].value, err_ => {
-        if (err_) {
-          console.error(err_.message);
-        }
-        callback();
-      });
-    });
-  });
+function addValues(hyperKey: regTypes.HKEY, commandKey: regTypes.HKEY, callback: Function) {
+  try {
+    Registry.setValueSZ(hyperKey, regParts[1].name, regParts[1].value);
+  } catch (error) {
+    console.error(error);
+  }
+  try {
+    Registry.setValueSZ(hyperKey, regParts[2].name, regParts[2].value);
+  } catch (err) {
+    console.error(err);
+  }
+  try {
+    Registry.setValueSZ(commandKey, regParts[0].name, regParts[0].value);
+  } catch (err_) {
+    console.error(err_);
+  }
+  callback();
 }
 
 export const add = (callback: Function) => {
-  const hyperKey = new Registry({hive: 'HKCU', key: regKey});
-  const commandKey = new Registry({
-    hive: 'HKCU',
-    key: `${regKey}\\${regParts[0].key}`
-  });
-
-  hyperKey.keyExists((error, exists) => {
-    if (error) {
-      console.error(error.message);
-    }
-    if (exists) {
-      commandKey.keyExists((err_, exists_) => {
-        if (err_) {
-          console.error(err_.message);
-        }
-        if (exists_) {
-          addValues(hyperKey, commandKey, callback);
-        } else {
-          commandKey.create(err => {
-            if (err) {
-              console.error(err.message);
-            }
-            addValues(hyperKey, commandKey, callback);
-          });
-        }
-      });
-    } else {
-      hyperKey.create(err => {
-        if (err) {
-          console.error(err.message);
-        }
-        commandKey.create(err_ => {
-          if (err_) {
-            console.error(err_.message);
-          }
-          addValues(hyperKey, commandKey, callback);
-        });
-      });
-    }
-  });
+  try {
+    const hyperKey =
+      Registry.openKey(Registry.HKCU, regKey, Registry.Access.ALL_ACCESS) ||
+      Registry.createKey(Registry.HKCU, regKey, Registry.Access.ALL_ACCESS);
+    const commandKey =
+      Registry.openKey(Registry.HKCU, `${regKey}\\${regParts[0].key}`, Registry.Access.ALL_ACCESS) ||
+      Registry.createKey(Registry.HKCU, `${regKey}\\${regParts[0].key}`, Registry.Access.ALL_ACCESS);
+    addValues(hyperKey, commandKey, callback);
+    Registry.closeKey(hyperKey);
+    Registry.closeKey(commandKey);
+  } catch (error) {
+    console.error(error);
+  }
 };
 
 export const remove = (callback: Function) => {
-  new Registry({hive: 'HKCU', key: regKey}).destroy(err => {
-    if (err) {
-      console.error(err.message);
-    }
-    callback();
-  });
+  try {
+    Registry.deleteTree(Registry.HKCU, regKey);
+  } catch (err) {
+    console.error(err);
+  }
+  callback();
 };

--- a/app/utils/cli-install.ts
+++ b/app/utils/cli-install.ts
@@ -5,11 +5,13 @@ import notify from '../notify';
 import {cliScriptPath, cliLinkPath} from '../config/paths';
 
 import * as regTypes from '../typings/native-reg';
-try {
-  // eslint-disable-next-line no-var, @typescript-eslint/no-var-requires
-  var Registry: typeof regTypes = require('native-reg');
-} catch (err) {
-  console.log(err);
+if (process.platform === 'win32') {
+  try {
+    // eslint-disable-next-line no-var, @typescript-eslint/no-var-requires
+    var Registry: typeof regTypes = require('native-reg');
+  } catch (err) {
+    console.error(err);
+  }
 }
 
 const readlink = pify(fs.readlink);

--- a/package.json
+++ b/package.json
@@ -88,7 +88,6 @@
     "@types/terser-webpack-plugin": "2.2.0",
     "@types/uuid": "7.0.0",
     "@types/webpack": "4.41.7",
-    "@types/winreg": "1.2.30",
     "@typescript-eslint/eslint-plugin": "2.21.0",
     "@typescript-eslint/parser": "2.21.0",
     "ava": "3.5.0",

